src/ucp/core/ucp_ep.h

```diff
@@ -194,14 +194,18 @@ typedef struct ucp_ep_config {
  * UCP_FEATURE_STREAM specific extention of the remote protocol layer endpoint
  */
 typedef struct ucp_ep_ext_stream {
-    /* ep which owns the extension */
-    ucp_ep_h                      ucp_ep;
     /* List entry in worker's EP list */
     ucs_list_link_t               list;
-    /* Queue of receive requests posted on the EP */
-    ucs_queue_head_t              reqs;
-    /* Queue of receive descriptors with data */
-    ucs_queue_head_t              data;
+    /* ep which owns the extension */
+    ucp_ep_h                      ucp_ep;
+    /* Total AM buffers length queued in @ref data, if > 0 @ref reqs is not valid */
+    size_t                        am_data_len;
```

Comment:
if the sole use of am_data_len is to distinguish between 'reqs' and 'data' it does not seem very efficient because it does not reduce memory usage. is it possible to use some ep flag instead? or maybe if the endpoint is queued on the worker it already means it has some data ?

Reply:
This counter allows to optimize recv_nb in place completion. In next PR, the rule will be a bit complicated - not just am_data_len > 0, but also need to check if we can complete with datatype entry alignment. In the future, most likely we will have `WAIT_ALL` flag, this also have to be checked.

Reply:
isn't it possible to use `desc->payload_offset` and `desc-length` to figure out the datatype entry alignment?

Reply:
I've implemented it a bit differently - i handle all possible rdescs and then decide complete the request or not. If it has not aligned data, the request wait next AM to fit alignment or request's buffer. I think to make prediction per rdesc is more costly than per request (taking into account WAIT_ALL, and using request on stack if it will be completed immediately).

Reply:
not sure this is the right approach, if there would be some generic datatype engine that "unpacks" receive descriptors, i guess it would be able to process only aligned amounts of data, maybe better to keep the remainder on the descriptor

Reply:
generic datatype will be completed with any amount of data. currently alignment is actual only for contig datatype.